### ISSUE TITLE
Linux/Mac; Context menu on Right Click Rocker Gesture "New Tab"

### ISSUE THREAD

**Issue by GSF1200S** [Robbendebiene/Gesturefy#164]
Recently a patch was accepted into Firefox that gives the about:config option "ui.context_menus.after_mouseup" to solve the Linux/Mac problem of context menus presenting when using right click as a mouse gesture.

Gesturefy and a competing extension FoxyGestures now both work fine with this option enabled in about:config. All trail gestures and wheel gestures work as they should.

The one exception is Rocker Gestures. Specifically holding left mouse and clicking right to open a new tab- a context menu displays once the newtab is opened. I think this might be impossible to fix given the limitation of webextensions on about:newtab, but I'm creating this bug to find out.

I should also note that on the Bugzilla bug page for context menu on mouse up one patch actually prevented this behavior. If you build dimimiem's patch into firefox a context menu will no present on newtab rocker gesture, BUT the context menu will show up with all trail gestures. Anyways, the current behavior is far better than that- I'm just mentioning in case it gives a clue as to why the bug I'm filing here occurs :) The mozilla bug is 1360278 (currently resolved)

**Comment by Robbendebiene** [comment]
Thanks for the detailed explanation and you are somehow right about the limitation.
The thing is that Rocker-Gestures should trigger on the second mousedown, therefore the mouseup (contextmenu) must be prevented after the action/gesture was fired. This normaly works well if you use "next tab" to go to a normal website but not for special pages like about:page (webextensions can not access them). Because of that it triggered on mouseup in older version of gesturefy. Since a lot of people complained about this behaviour I changed it, which is indeed the better solution for most of the actions except things like "new tab".
But even if I would change the behaviour back/add a setting to change it, this problem would still occur if you "hold right" > "click left" and release the right button on the new "tab page". So at the end I can't completely guarantee the contextmenu supression, because it depends on the moment, the user releases the right mouse button.

> I should also note that on the Bugzilla bug page for context menu on mouse up one patch actually prevented this behavior. If you build dimimiem's patch into firefox a context menu will no present on newtab rocker gesture

This is interesting and strange at once.  Maybe the actual event was still called on mousedown, but the contextmenu was shown on mouseup. I have to admit, that I don't have good explanation for that :)

**Comment by knoajp** [comment]
I set the Rocker gesture "Right mouse click" [Left to Right] for opening a new tab and having no issue for months or so. (it seems to be strange after I read the comments above!?)
But from recent weeks or a month at most, it happens to popup contextmenu on the opened tab. (regardless of about:blank, about:home or any web pages)
It sometimes happens, and it might depend on the timing of the mouseup of the right click.

How to reproduce:
1. set the Rocker gesture "Right mouse click" [Left to Right] for opening a new tab
2. visit some gesturable web page
3. fire the Rocker gesture "Right mouse click" [Left to Right]
4. contextmenu happens to popup on the new tab (depends on the timing of the mouseup of the right click)

Does it cause the issue?
https://github.com/Robbendebiene/Gesturefy/commit/160500646c18cdbc56c8899caa7cbd7a3bda65e5
(I don't have any clues but I just found the commit which has 'contextmenu' on recent versions)

It looks similar but my case is from recent weeks.
https://github.com/Robbendebiene/Gesturefy/issues/529

It could be occurred by Firefox's update, but I don't have any idea.
(macOS Catalina 10.15.7, Firefox Beta 87.0 and now updated to 88.0, Gesturefy 3.1.1)

**Comment by Robbendebiene** [comment]
@knoajp If it only occurs sometimes then this is probably the same issue as #529. It's a race condition where the new tab is opened before the context menu event fired. I didn't come up with a **good** solution yet.

**Comment by knoajp** [comment]
Thanks for the quick reply.
So the mystery is that the issue just started a few weeks ago for me... 🙃
I wonder if Firefox's event handler has slightly changed or something.

**Comment by Robbendebiene** [comment]
I guess there could be a lot of reasons. It might also be caused by the the operating system. If you have any updates on this (maybe you found a way to reproduce it reliable) please leave further comments in the other issue :)

**Comment by knoajp** [comment]
You are an insightful doctor!!
I could find the very condition to increase the chance for popping up very much.

The condition is:
The macro app "Keyboard Maestro" engine is observing the focused window change for triggering macros.

If I remove the observation, the popup rarely wins the "race". (quite a timing issue)
I did make the condition recently, and I removed it now.

It's beyond my idea.
Thank you very much.
